cmds/core/timeout/main_test.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bytes"
+	"errors"
+	"os/exec"
+	"testing"
+
+	"github.com/u-root/u-root/pkg/testutil"
+)
+
+// TestOK for now just runs a simple successful test with 0 args or more than one arg.
+func TestOK(t *testing.T) {
+	var tests = []struct {
+		cmd cmd
+		err error
+	}{
+		{cmd: cmd{args: []string{}}, err: errNoArgs},
+		{cmd: cmd{args: []string{"date"}}, err: errBadDuration},
+		{cmd: cmd{args: []string{"date"}, timeout: "wh"}, err: errBadDuration},
+	}
+
+	for _, v := range tests {
+		t.Logf("Run %v", v.cmd)
+		if err := v.cmd.run(); !errors.Is(err, v.err) {
+			t.Errorf("run %v: got %v, want %v", v.cmd, err, v.err)
+		}
+	}
+}
+
+func TestRun(t *testing.T) {
+	if _, err := exec.LookPath("sleep"); err != nil {
+		t.Skipf("Skipping this test as sleep is not in the path")
+	}
+
+	var tests = []struct {
+		cmd cmd
+		ok  bool
+	}{
+		{cmd: cmd{args: []string{"sleep", "4"}, timeout: "2m"}, ok: true},
+		{cmd: cmd{args: []string{"sleep", "30"}, timeout: "1s"}},
+	}
+	for _, v := range tests {
+		t.Logf("Run %v", v.cmd)
+		// Return errors from running are not guaranteed.
+		// Just see if succeeded or not.
+		if err := v.cmd.run(); err == nil != v.ok {
+			t.Errorf("run %v: got %v, want %v", v.cmd, err == nil, v.ok)
+		}
+	}
+
+}
+
+// Test real execution. Why do this if we covered above?
```

Comment:
(optional) I wonder what happened before such that the program broke but coverage was fine?  Was the problem with flag parsing?

It looks like this function is testing the actual flags on the command line ... I think you may be able to get the same test coverage without the performance hit of forking a new process to run the test with testutil.Command if you follow a pattern like this .... you can then isolate and test the 'cmd' struct to check that 'timeout' and 'args' is populated properly directly from the arguments you pass in during the test.

```
func main() {
    c, err := newCMD(os.Args, os.Stdin, os.Stdout, os.Stderr)
    if err != nil {
        log.Fatal(err)
    }
    if err := c.run(); err != nil {
        log.Fatalf("timeout(%v):%v", c.timeout, err)
    }
}

func newCMD(args []string, stdin, stdout, stderr *os.File) (*cmd, error) {
    c := new(cmd)

    // Create a new flag set and populate it from the command line arguments.
    flagSet := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
    flagSet.DurationVar(&c.timeout, "t", 30 * time.Second, "Timeout for command")

    if err := flagSet.Parse(args); err != nil {
        return nil, fmt.Errorf("error parsing flags: %v", err)
    }
    c.args = flagSet.Args()

    c.in, c.out, c.err = stdin, stdout, stderr

    return c, nil
}
```

Reply:
It was a different program, I'll make that comment clearer. w.r.t. to the (optional) section, I'm a bit worried about 
presenting that to the wider users of this code -- it seems a bit much to me.

Reply:
I don't think I want to test the flag package here. If the sequence
-t=1m somecommand -t=true
fails then it is the flag package breaking and I have a feeling somebody else would know that first. 

It's a fine line, I know, but in this case I'd rather not cross it. Thanks!

Reply:
Agreed, it's a fine line :) I was thinking more that we'd be able to change the underlying flag parsing implementation and have good test coverage, but it seems unlikely that we'd move away from the 'flag' package, so this is fine with me :)